src/rdb.c

```diff
@@ -2436,35 +2360,30 @@ int rdbSaveToSlavesSockets(rdbSaveInfo *rsi) {
 
     if (server.aof_child_pid != -1 || server.rdb_child_pid != -1) return C_ERR;
 
-    /* Before to fork, create a pipe that will be used in order to
-     * send back to the parent the IDs of the slaves that successfully
-     * received all the writes. */
+    /* Even if the previous fork child exited, don't start a new one until we
+     * drained the pipe. */
+    if (server.rdb_pipe_conns) return C_ERR;
+
+    /* Before to fork, create a pipe that is used to transfer the rdb bytes to
+     * the parant, we can't let it write directly to the sockets, since in case
+     * of TLS we must let the parent handle a contineous TLS state when the
+     * child terminates and parent takes over. */
     if (pipe(pipefds) == -1) return C_ERR;
-    server.rdb_pipe_read_result_from_child = pipefds[0];
-    server.rdb_pipe_write_result_to_parent = pipefds[1];
+    server.rdb_pipe_read = pipefds[0];
+    server.rdb_pipe_write = pipefds[1];
+    anetNonBlock(NULL, server.rdb_pipe_read);
 
-    /* Collect the file descriptors of the slaves we want to transfer
+    /* Collect the connections of the replicas we want to transfer
      * the RDB to, which are i WAIT_BGSAVE_START state. */
-    fds = zmalloc(sizeof(int)*listLength(server.slaves));
-    /* We also allocate an array of corresponding client IDs. This will
-     * be useful for the child process in order to build the report
-     * (sent via unix pipe) that will be sent to the parent. */
-    clientids = zmalloc(sizeof(uint64_t)*listLength(server.slaves));
-    numfds = 0;
-
+    server.rdb_pipe_conns = zmalloc(sizeof(connection *)*listLength(server.slaves));
```

Comment:
At this point the replicas connections have two owners, the list itself, and the pipe connections. This is in theory a potential source of problems in case the replica is removed from one place but remains in the piped connection. I guess this is not possible with the current code, but just wanted to warn about making sure the code is structured to guarantee this property.

Reply:
@antirez sorry for the late response.
I did nullify these in unlinkClient in case a replica disconnects while rdb fork is still active. And I did add a test for coverage in which some/all replicas drop during transfer. 
I know it's not great, but I think we can live with it, if you have a better idea let me know.

Reply:
Hello Oran, yes, this makes sense indeed. I guess that the only thing we can do better than that is to use some kind of flag and asserts() in freeClient(), but it's a small detail that we can fix later. Thank you.